drivers/sensor/bosch/bmp180/bmp180.c

```diff
@@ -0,0 +1,487 @@
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/pm/device.h>
+#include <math.h>
+
+#include "bmp180.h"
+
+LOG_MODULE_REGISTER(BMP180, CONFIG_SENSOR_LOG_LEVEL);
+
+static inline int bmp180_bus_check(const struct device *dev)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return cfg->bus_io->check(&cfg->bus);
+}
+
+static inline int bmp180_reg_read(const struct device *dev,
+				  uint8_t start, uint8_t *buf, int size)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return cfg->bus_io->read(&cfg->bus, start, buf, size);
+}
+
+static inline int bmp180_reg_write(const struct device *dev, uint8_t reg,
+				   uint8_t val)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return cfg->bus_io->write(&cfg->bus, reg, val);
+}
+
+int bmp180_reg_field_update(const struct device *dev,
+			    uint8_t reg,
+			    uint8_t mask,
+			    uint8_t val)
+{
+	int rc = 0;
+	uint8_t old_value, new_value;
+	const struct bmp180_config *cfg = dev->config;
+
+	rc = cfg->bus_io->read(&cfg->bus, reg, &old_value, 1);
+	if (rc != 0) {
+		return rc;
+	}
+
+	new_value = (old_value & ~mask) | (val & mask);
+	if (new_value == old_value) {
+		return 0;
+	}
+
+	return cfg->bus_io->write(&cfg->bus, reg, new_value);
+}
+
+#ifdef CONFIG_BMP180_OSR_RUNTIME
+static int bmp180_attr_set_oversampling(const struct device *dev,
+					enum sensor_channel chan,
+					uint16_t val)
+{
+	struct bmp180_data *data = dev->data;
+
+	/* Value must be a positive value 0-3 */
+	if(val > 3) {
+		return -EINVAL;
+	}
+
+	if (chan == SENSOR_CHAN_PRESS) {
+	     data->osr_pressure = val;
+	}
+	else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+#endif
+
+static int bmp180_attr_set(const struct device *dev,
+			   enum sensor_channel chan,
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
+{
+	int ret;
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(dev, &state);
+	if (state != PM_DEVICE_STATE_ACTIVE) {
+		return -EBUSY;
+	}
+#endif
+
+	switch (attr) {
+#ifdef CONFIG_BMP180_OSR_RUNTIME
+	case SENSOR_ATTR_OVERSAMPLING:
+		ret = bmp180_attr_set_oversampling(dev, chan, val->val1);
+		break;
+#endif
+
+	default:
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+
+static int read_raw_temperature(const struct device *dev)
+{
+	int ret;
+	bool wait = true;
+	uint8_t regControl, msb, lsb;
+	struct bmp180_data *data = dev->data;
+
+	/* send temperature measurement command */
+	ret = bmp180_reg_write(dev, BMP180_REG_MEAS_CTRL, BMP180_CMD_GET_TEMPERATURE);
+	if (ret) {
+		return ret;
+	}
+
+	/* for the first while read delay is 3000+1000 */
+	k_sleep(K_MSEC(3));
+
+	while(wait) {
+		k_sleep(K_MSEC(1));
+		ret = bmp180_reg_read(dev, BMP180_REG_MEAS_CTRL, &regControl, 1);
+		if (ret) {
+			return ret;
+		}
+
+		/* bit 5 is 1 until data registers are ready */ 
+		if ((regControl & BMP180_STATUS_CMD_RDY)==0) {
+			wait = false;
+		}
+	} 
+
+	ret = bmp180_reg_read(dev, BMP180_REG_MSB, &msb, 1);
```

Comment:
Not sure if the ic would support it but maybe its possible to read 2 bytes at once?

Reply:
There is a note in the data-sheet:
After conversion, data registers can be read out in any sequence (i.e. MSB first or LSB first) Using a burst read is not mandatory.
So I use dedicated read for the registers.

Reply:
I update the code using burst read for temp & pressure  and it works fine. Applied
struct reg16_t {
        uint8_t msb;
        uint8_t lsb;
    } reg16;
ret = bmp180_reg_read(dev, BMP180_REG_MSB, (uint8_t *)&reg16, sizeof(struct reg16_t));
    if (ret) {
        return ret;
    }
data->raw_temp = ((int32_t)reg16.msb<<8) | (int32_t)reg16.lsb;

Reply:
I'd say just use a `uint8_t buffer[2]` no need for a dedicated structure. You can then use `sys_get_be16` to convert the buffer to the uint16_t value, less self-written bit manipulation. https://docs.zephyrproject.org/latest/doxygen/html/sys_2byteorder_8h.html#a91fae25b12126c2b35d240f56866184d

Reply:
My intention with the structure is to describe / make it visible what it is.
uint8_t reg16[sizeof(uint16t)] would hide it.  

Need to cast the sys_get_be16(). 
temp = (int16_t)(sys_get_be16((uint16_t)(reg16))) ;

Reply:
uint8_t reg16/24[] used. Resolved